File: apps/api/app/routers/roadmaps.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.schemas.arm import ArmRow
from app.schemas.roadmap import MilestonePublic, RoadmapSummary
from app.services import roadmap as roadmap_service

router = APIRouter(prefix="/roadmaps", tags=["roadmaps"])


def _summary(db: Session, r) -> RoadmapSummary:
    return RoadmapSummary(
        id=r.id,
        title=r.title,
        goal_summary=r.goal_summary,
        position=r.position,
        arm=ArmRow.model_validate(r.arm),
        milestones=[MilestonePublic.model_validate(m) for m in roadmap_service.list_milestones(db, r)],
    )


@router.get("", response_model=list[RoadmapSummary])
def list_roadmaps(arm: str | None = None, db: Session = Depends(get_db)):
    return [_summary(db, r) for r in roadmap_service.list_published_roadmaps(db, arm_slug=arm)]

File: apps/web/src/lib/api/roadmaps.ts
import { apiFetch } from "./core";
import type { Arm } from "./shared";

export type MilestoneStatus = "planned" | "in_progress" | "done";

export type MilestonePublic = {
  id: string;
  title: string;
  description: string | null;
  status: MilestoneStatus;
  position: number;
};

export type RoadmapSummary = {
  id: string;
  title: string;
  goal_summary: string | null;
  position: number;
  arm: Arm;
  milestones: MilestonePublic[];
};

export const roadmapApi = {
  list: (armSlug?: string) => apiFetch<RoadmapSummary[]>(`/roadmaps${armSlug ? `?arm=${armSlug}` : ""}`),
};
